slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/DbSlashingProtection.java

```diff
@@ -36,12 +48,40 @@ public boolean maySignAttestation(
 
   @Override
   public boolean maySignBlock(
-      final String publicKey, final Bytes signingRoot, final UInt64 blockSlot) {
-    return true;
+      final Bytes publicKey, final Bytes signingRoot, final UInt64 blockSlot) {
+    return jdbi.inTransaction(
+        h -> {
+          final List<Validator> validators =
+              validatorsDao.retrieveValidators(h, List.of(publicKey));
+          final Optional<Long> validatorId = validators.stream().findFirst().map(Validator::getId);
+          if (validatorId.isEmpty()) {
+            return false;
+          } else {
+            final long id = validatorId.get();
+            final Optional<SignedBlock> existingBlock =
+                signedBlocksDao.findExistingBlock(h, id, blockSlot);
+
+            // same slot and signing_root is allowed for broadcasting previously signed block
+            // otherwise if slot and different signing_root then this is a double block proposal
+            final boolean isValid =
+                existingBlock.map(block -> block.getSigningRoot().equals(signingRoot)).orElse(true);
```

Comment:
orElse true or false??

Reply:
so if existing block is empty, it is valid!

Reply:
Should be orElse true. What it's saying is if we haven't signed this block before then it's valid to sign it. We are only trying to avoid signing the same block for the slot and public key more than once.